File: src/main/java/dev/verloren/midnight/lexer/CompactTokenTypes.java
package dev.verloren.midnight.lexer;


import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IElementType;

/**
 * Static registry of all lexical token types recognised in Compact source code.
 *
 * <p>Contains definitions for:
 * <ul>
 *   <li><b>Keywords:</b> Compact language keywords such as {@code circuit}, {@code witness}, {@code ledger}, {@code struct}, {@code enum}, {@code contract}.</li>
 *   <li><b>Builtin Types:</b> Primitive type keywords such as {@code Boolean}, {@code Field}, {@code Uint}, {@code Bytes}, {@code Vector}.</li>
 *   <li><b>Literals:</b> Boolean, numeric (decimal, hex, binary, octal), version, and string literals.</li>
 *   <li><b>Operators & Delimiters:</b> Arithmetic, relational, logical, assignment, and punctuation tokens.</li>
 *   <li><b>Comments & Whitespace:</b> Line and block comments, whitespace, and lexical error tokens.</li>
 * </ul>
 * </p>
 */
public final class CompactTokenTypes {

  public static final IElementType INVALID_VERSION = new CompactTokenType("INVALID_VERSION");


  // =========================================================
  // Keywords
  // =========================================================
  public static final IElementType PRAGMA = new CompactTokenType("PRAGMA");
  public static final IElementType EXPORT = new CompactTokenType("EXPORT");
  public static final IElementType FROM = new CompactTokenType("FROM");
  public static final IElementType IMPORT = new CompactTokenType("IMPORT");
  public static final IElementType MODULE = new CompactTokenType("MODULE");
  public static final IElementType PREFIX = new CompactTokenType("PREFIX");
  public static final IElementType ASSERT = new CompactTokenType("ASSERT");
  public static final IElementType AS = new CompactTokenType("AS");
  public static final IElementType CIRCUIT = new CompactTokenType("CIRCUIT");
  public static final IElementType CONST = new CompactTokenType("CONST");
  public static final IElementType CONSTRUCTOR = new CompactTokenType("CONSTRUCTOR");
  public static final IElementType CONTRACT = new CompactTokenType("CONTRACT");
  public static final IElementType DEFAULT = new CompactTokenType("DEFAULT");
  public static final IElementType DISCLOSE = new CompactTokenType("DISCLOSE");
  public static final IElementType ELSE = new CompactTokenType("ELSE");
  public static final IElementType ENUM = new CompactTokenType("ENUM");
  public static final IElementType FOLD = new CompactTokenType("FOLD");
  public static final IElementType FOR = new CompactTokenType("FOR");
  public static final IElementType IF = new CompactTokenType("IF");
  public static final IElementType INCLUDE = new CompactTokenType("INCLUDE");
  public static final IElementType LEDGER = new CompactTokenType("LEDGER");
  public static final IElementType MAP = new CompactTokenType("MAP");
  public static final IElementType NEW = new CompactTokenType("NEW");
  public static final IElementType OF = new CompactTokenType("OF");
  public static final IElementType PAD = new CompactTokenType("PAD");
  public static final IElementType PURE = new CompactTokenType("PURE");
  public static final IElementType RETURN = new CompactTokenType("RETURN");
  public static final IElementType SEALED = new CompactTokenType("SEALED");
  public static final IElementType SLICE = new CompactTokenType("SLICE");
  public static final IElementType STRUCT = new CompactTokenType("STRUCT");
  public static final IElementType TYPE = new CompactTokenType("TYPE");
  public static final IElementType WITNESS = new CompactTokenType("WITNESS");
  public static final IElementType EMIT = new CompactTokenType("EMIT");
  public static final IElementType IMPLEMENTS = new CompactTokenType("IMPLEMENTS");
  public static final IElementType EXTERNAL = new CompactTokenType("EXTERNAL");
  public static final IElementType HASH = new CompactTokenType("HASH");

  // =========================================================
  // Built-in Types
  // =========================================================

  public static final IElementType BOOLEAN_TYPE = new CompactTokenType("BOOLEAN_TYPE");
  public static final IElementType BYTES_TYPE = new CompactTokenType("BYTES_TYPE");
  public static final IElementType FIELD_TYPE = new CompactTokenType("FIELD_TYPE");
  public static final IElementType OPAQUE_TYPE = new CompactTokenType("OPAQUE_TYPE");
  public static final IElementType UINT_TYPE = new CompactTokenType("UINT_TYPE");
  public static final IElementType VECTOR_TYPE = new CompactTokenType("VECTOR_TYPE");
  public static final IElementType JUBJUB_SCALAR_TYPE = new CompactTokenType("JUBJUB_SCALAR_TYPE");
  public static final IElementType SECP256K1_BASE_TYPE = new CompactTokenType("SECP256K1_BASE_TYPE");
  public static final IElementType SECP256K1_SCALAR_TYPE = new CompactTokenType("SECP256K1_SCALAR_TYPE");

// =========================================================
// Literals
// =========================================================

  public static final IElementType TRUE = new CompactTokenType("TRUE");
  public static final IElementType FALSE = new CompactTokenType("FALSE");

  public static final IElementType VERSION_LITERAL = new CompactTokenType("VERSION_LITERAL");
  public static final IElementType DECIMAL_LITERAL = new CompactTokenType("DECIMAL_LITERAL");
  public static final IElementType BINARY_LITERAL = new CompactTokenType("BINARY_LITERAL");
  public static final IElementType OCTAL_LITERAL = new CompactTokenType("OCTAL_LITERAL");
  public static final IElementType HEX_LITERAL = new CompactTokenType("HEX_LITERAL");
  public static final IElementType STRING_LITERAL = new CompactTokenType("STRING_LITERAL");

  public static final IElementType IDENTIFIER = new CompactTokenType("IDENTIFIER");

  // =========================================================
  // Operators
  // =========================================================

  public static final IElementType ASSIGN = new CompactTokenType("ASSIGN");

  public static final IElementType PLUS_ASSIGN = new CompactTokenType("PLUS_ASSIGN");
  public static final IElementType MINUS_ASSIGN = new CompactTokenType("MINUS_ASSIGN");

  public static final IElementType PLUS = new CompactTokenType("PLUS");
  public static final IElementType MINUS = new CompactTokenType("MINUS");
  public static final IElementType STAR = new CompactTokenType("STAR");
  public static final IElementType SLASH = new CompactTokenType("SLASH");
  public static final IElementType PERCENT = new CompactTokenType("PERCENT");

  public static final IElementType EQEQ = new CompactTokenType("EQEQ");
  public static final IElementType NEQ = new CompactTokenType("NEQ");

  public static final IElementType LT = new CompactTokenType("LT");
  public static final IElementType LTE = new CompactTokenType("LTE");
  public static final IElementType GT = new CompactTokenType("GT");
  public static final IElementType GTE = new CompactTokenType("GTE");

  public static final IElementType ARROW = new CompactTokenType("ARROW");

  public static final IElementType NOT = new CompactTokenType("NOT");
  public static final IElementType ANDAND = new CompactTokenType("ANDAND");
  public static final IElementType OROR = new CompactTokenType("OROR");

  public static final IElementType RANGE = new CompactTokenType("RANGE");
  public static final IElementType DOT = new CompactTokenType("DOT");
  public static final IElementType SPREAD = new CompactTokenType("SPREAD");
  // =========================================================
// Delimiters
// =========================================================

  public static final IElementType LPAREN = new CompactTokenType("LPAREN");
  public static final IElementType RPAREN = new CompactTokenType("RPAREN");

  public static final IElementType LBRACE = new CompactTokenType("LBRACE");
  public static final IElementType RBRACE = new CompactTokenType("RBRACE");

  public static final IElementType LBRACKET = new CompactTokenType("LBRACKET");
  public static final IElementType RBRACKET = new CompactTokenType("RBRACKET");

  public static final IElementType COMMA = new CompactTokenType("COMMA");
  public static final IElementType COLON = new CompactTokenType("COLON");
  public static final IElementType SEMICOLON = new CompactTokenType("SEMICOLON");

// =========================================================
// Comments
// =========================================================

  public static final IElementType LINE_COMMENT = new CompactTokenType("LINE_COMMENT");
  public static final IElementType BLOCK_COMMENT = new CompactTokenType("BLOCK_COMMENT");

// =========================================================
// Whitespace & Special
// =========================================================

  public static final IElementType WHITE_SPACE = TokenType.WHITE_SPACE;
  public static final IElementType BAD_CHARACTER = TokenType.BAD_CHARACTER;
  public static final IElementType UNTERMINATED_BLOCK_COMMENT =
          new CompactTokenType("UNTERMINATED_BLOCK_COMMENT");

  public static final IElementType UNTERMINATED_STRING =
          new CompactTokenType("UNTERMINATED_STRING");

  public static final IElementType LET = new CompactTokenType("LET");
  public static final IElementType QUESTION = new CompactTokenType("QUESTION");

  private CompactTokenTypes() {
  }
}

File: src/main/kotlin/dev/verloren/midnight/lexer/CompactTokenTypes.kt
package dev.verloren.midnight.lexer

import com.intellij.psi.TokenType
import com.intellij.psi.tree.IElementType

object CompactTokenTypes {

    // =========================================================
    // Keywords
    // =========================================================

    @JvmField
    val EXPORT = CompactTokenType("EXPORT")

    @JvmField
    val FROM = CompactTokenType("FROM")

    @JvmField
    val IMPORT = CompactTokenType("IMPORT")

    @JvmField
    val MODULE = CompactTokenType("MODULE")

    @JvmField
    val PREFIX = CompactTokenType("PREFIX")

    @JvmField
    val ASSERT = CompactTokenType("ASSERT")

    @JvmField
    val AS = CompactTokenType("AS")

    @JvmField
    val CIRCUIT = CompactTokenType("CIRCUIT")

    @JvmField
    val CONST = CompactTokenType("CONST")

    @JvmField
    val CONSTRUCTOR = CompactTokenType("CONSTRUCTOR")

    @JvmField
    val CONTRACT = CompactTokenType("CONTRACT")

    @JvmField
    val DEFAULT = CompactTokenType("DEFAULT")

    @JvmField
    val DISCLOSE = CompactTokenType("DISCLOSE")

    @JvmField
    val ELSE = CompactTokenType("ELSE")

    @JvmField
    val ENUM = CompactTokenType("ENUM")

    @JvmField
    val FOLD = CompactTokenType("FOLD")

    @JvmField
    val FOR = CompactTokenType("FOR")

    @JvmField
    val IF = CompactTokenType("IF")

    @JvmField
    val INCLUDE = CompactTokenType("INCLUDE")

    @JvmField
    val LEDGER = CompactTokenType("LEDGER")

    @JvmField
    val MAP = CompactTokenType("MAP")

    @JvmField
    val NEW = CompactTokenType("NEW")

    @JvmField
    val OF = CompactTokenType("OF")

    @JvmField
    val PAD = CompactTokenType("PAD")

    @JvmField
    val PRAGMA = CompactTokenType("PRAGMA")

    @JvmField
    val PURE = CompactTokenType("PURE")

    @JvmField
    val RETURN = CompactTokenType("RETURN")

    @JvmField
    val SEALED = CompactTokenType("SEALED")

    @JvmField
    val SLICE = CompactTokenType("SLICE")

    @JvmField
    val STRUCT = CompactTokenType("STRUCT")

    @JvmField
    val TYPE = CompactTokenType("TYPE")

    @JvmField
    val WITNESS = CompactTokenType("WITNESS")

    // =========================================================
    // Built-in Types
    // =========================================================

    @JvmField
    val BOOLEAN_TYPE = CompactTokenType("BOOLEAN_TYPE")

    @JvmField
    val BYTES_TYPE = CompactTokenType("BYTES_TYPE")

    @JvmField
    val FIELD_TYPE = CompactTokenType("FIELD_TYPE")

    @JvmField
    val OPAQUE_TYPE = CompactTokenType("OPAQUE_TYPE")

    @JvmField
    val UINT_TYPE = CompactTokenType("UINT_TYPE")

    @JvmField
    val VECTOR_TYPE = CompactTokenType("VECTOR_TYPE")

    // =========================================================
    // Literals
    // =========================================================

    @JvmField
    val TRUE = CompactTokenType("TRUE")

    @JvmField
    val FALSE = CompactTokenType("FALSE")

    @JvmField
    val VERSION_LITERAL = CompactTokenType("VERSION_LITERAL")

    @JvmField
    val DECIMAL_LITERAL = CompactTokenType("DECIMAL_LITERAL")

    @JvmField
    val BINARY_LITERAL = CompactTokenType("BINARY_LITERAL")

    @JvmField
    val OCTAL_LITERAL = CompactTokenType("OCTAL_LITERAL")

    @JvmField
    val HEX_LITERAL = CompactTokenType("HEX_LITERAL")


    @JvmField
    val STRING_LITERAL = CompactTokenType("STRING_LITERAL")

    @JvmField
    val IDENTIFIER = CompactTokenType("IDENTIFIER")

    // =========================================================
    // Reserved Keywords
    // =========================================================

    @JvmField
    val RESERVED_KEYWORD = CompactTokenType("RESERVED_KEYWORD")

    // =========================================================
    // Operators
    // =========================================================

    @JvmField
    val ASSIGN = CompactTokenType("=")

    @JvmField
    val PLUS_ASSIGN = CompactTokenType("+=")

    @JvmField
    val MINUS_ASSIGN = CompactTokenType("-=")

    @JvmField
    val PLUS = CompactTokenType("+")

    @JvmField
    val MINUS = CompactTokenType("-")

    @JvmField
    val STAR = CompactTokenType("*")

    @JvmField
    val SLASH = CompactTokenType("/")

    @JvmField
    val PERCENT = CompactTokenType("%")

    @JvmField
    val EQEQ = CompactTokenType("==")

    @JvmField
    val NEQ = CompactTokenType("!=")

    @JvmField
    val LT = CompactTokenType("<")

    @JvmField
    val LTE = CompactTokenType("<=")

    @JvmField
    val GT = CompactTokenType(">")

    @JvmField
    val GTE = CompactTokenType(">=")

    @JvmField
    val ARROW = CompactTokenType("=>")

    @JvmField
    val NOT = CompactTokenType("!")

    @JvmField
    val ANDAND = CompactTokenType("&&")

    @JvmField
    val OROR = CompactTokenType("||")

    @JvmField
    val RANGE = CompactTokenType("..")

    @JvmField
    val DOT = CompactTokenType(".")

    @JvmField
    val SPREAD = CompactTokenType("...")

    @JvmField
    val QUESTION = CompactTokenType("?")

    @JvmField
    val COLON = CompactTokenType(":")

    // =========================================================
    // Delimiters
    // =========================================================

    @JvmField
    val LPAREN = CompactTokenType("(")

    @JvmField
    val RPAREN = CompactTokenType(")")

    @JvmField
    val LBRACE = CompactTokenType("{")

    @JvmField
    val RBRACE = CompactTokenType("}")

    @JvmField
    val LBRACKET = CompactTokenType("[")

    @JvmField
    val RBRACKET = CompactTokenType("]")

    @JvmField
    val COMMA = CompactTokenType(",")

    @JvmField
    val SEMICOLON = CompactTokenType(";")

    @JvmField
    val HASH = CompactTokenType("#")

    // =========================================================
    // Comments
    // =========================================================

    @JvmField
    val LINE_COMMENT = CompactTokenType("LINE_COMMENT")

    @JvmField
    val BLOCK_COMMENT = CompactTokenType("BLOCK_COMMENT")

    // =========================================================
    // Whitespace & Special
    // =========================================================

    @JvmField
    val WHITE_SPACE: IElementType = TokenType.WHITE_SPACE

    @JvmField
    val BAD_CHARACTER: IElementType = TokenType.BAD_CHARACTER

    @JvmField
    val UNTERMINATED_STRING = CompactTokenType("UNTERMINATED_STRING")

    @JvmField
    val UNTERMINATED_BLOCK_COMMENT =
        CompactTokenType("UNTERMINATED_BLOCK_COMMENT")
}
